cirq/protocols/apply_mixture_protocol.py

```diff
@@ -0,0 +1,309 @@
+            tensor, available workspace, and left and right axes to operate on.
+            The attributes of this object will be mutated as part of computing
+            the result.
+        default: What should be returned if `val` doesn't have a mixture. If
+            not specified, a TypeError is raised instead of returning a default
+            value.
+
+    Returns:
+        If the receiving object is not able to apply a mixture,
+        the specified default value is returned (or a TypeError is raised). If
+        this occurs, then `target_tensor` should not have been mutated.
+
+        If the receiving object was able to work inline, directly
+        mutating `target_tensor` it will return `target_tensor`. The caller is
+        responsible for checking if the result is `target_tensor`.
+
+        If the receiving object wrote its output over `out_buffer`, the
+        result will be `out_buffer`. The caller is responsible for
+        checking if the result is `out_buffer` (and e.g. swapping
+        the buffer for the target tensor before the next call).
+
+        Note that it is an error for the return object to be either of the
+        auxiliary buffers, and the method will raise an AssertionError if
+        this contract is violated.
+
+        The receiving object may also write its output over a new buffer
+        that it created, in which case that new array is returned.
+
+    Raises:
+        TypeError: `val` doesn't have a mixture and `default` wasn't specified.
+        ValueError: Different left and right shapes of `args.target_tensor`
+            selected by `left_axes` and `right_axes` or `qid_shape(val)` doesn't
+            equal the left and right shapes.
+        AssertionError: `_apply_mixture_` returned an auxiliary buffer.
+    """
+
+    # Verify that val has the same qid shape as the selected axes of the density
+    # matrix tensor.
+    val_qid_shape = qid_shape_protocol.qid_shape(val,
+                                                 (2,) * len(args.left_axes))
+    left_shape = tuple(args.target_tensor.shape[i] for i in args.left_axes)
+    right_shape = tuple(args.target_tensor.shape[i] for i in args.right_axes)
+    if left_shape != right_shape:
+        raise ValueError('Invalid target_tensor shape or selected axes. '
+                         'The selected left and right shape of target_tensor '
+                         'are not equal. Got {!r} and {!r}.'.format(
+                             left_shape, right_shape))
+    if val_qid_shape != left_shape:
+        raise ValueError('Invalid mixture qid shape is not equal to the '
+                         'selected left and right shape of target_tensor. '
+                         'Got {!r} but expected {!r}.'.format(
+                             val_qid_shape, left_shape))
+
+    # Check if the specialized method is present.
+    func = getattr(val, '_apply_mixture_', None)
```

Comment:
Rewrite this method in the style of `cirq.apply_unitary`, i.e. an explicit list of strategies which is iterated over until one succeeds with each strategy being a method that is self-contained. So the general progression would be:

1. Validate input.
2. Make strategies list.
3. Iterate over strategies trying to get a result.
4. Fail or return default value.

Reply:
Outlined the procedure, let me know what you think